### PR TITLE
Assignment3

### PR DESCRIPTION
**Pull request by jstanley6**
Finished Assignment 3

### CODE REVIEW

**Comment by Malcolm-Shepherd** [comment]
## Assignment

For this assignment do the following:

- Create *two* interfaces. One to represent the `JokeService` and one interface to represent outputting the joke to the screen (`Console.WriteLine`). ✔
- The `JokeService` will need to have the interface applied to it. ❌
- Create an implementation of the output interface that writes out the joke to the console. ✔
- Implement the `Jester` class. It should take in both interfaces as dependencies. These dependencies should be null checked. ✔
- The `Jester` class `TellJoke()` method should retrieve a joke from the `JokeService`. If the joke contains "Chuck Norris", skip it and get another. The joke should be written to the output dependency. ✔
- Unit test the Jester class. Code coverage should be above 90% for this class. Moq will make this significantly easier. ✔

## Fundamentals

- Be sure you enable:
  - Nullability for all projects  ❌
  - Set `LangVersion` and the `TargetFramework` to the latest released versions available (preview versions optional) ❌
  - Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ❌
- For this assignment, always use `Assert.AreEqual<T>()` (the generic version)  ❌
- **Ensure there are no errors or warnings (including code analysis warnings)** ✔
- **All of the above should be unit tested** ❌
- **Choose simplicity over complexity** ✔

**Comment by Malcolm-Shepherd** [comment]
## The JokeService will need to have the interface applied to it 
You implemented the Output interface in the output class, but forgot to do the same for 'IJokeService' and 'JokeService'
## Be sure you enable (nullability, LangVersion, TargetFramework, etc)
Make sure you add the following to each of your .csproj files in the future
'
   <TargetFramework>net7.0</TargetFramework>
	<LangVersion>11.0</LangVersion>
	<Nullable>enable</Nullable>  
	<EnableNETAnalyzers>true</EnableNETAnalyzers>
	<TreatWarningsAsErrors>true</TreatWarningsAsErrors><CodeAnalysisTreatWarningsAsErrors>true</CodeAnalysisTreatWarningsAsErrors>
	'
## For this assignment, always use Assert.AreEqual<T>() (the generic version)
They wanted us to specify the type when using Assert.AreEqual. In your 'InitializeJester_Pass' test you should have used Assert.AreEqual<JokeService> instead

File: Assignment3/CanHazFunny.Tests/JesterTests.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+            Assert.IsNotNull(testJester);
+            string joke = testJester.TellJoke();
+            Assert.IsFalse(String.IsNullOrEmpty(joke));
+
+
+        }
+        
+    }
+}
+
```
Comment (Bhopss152) [inline Assignment3/CanHazFunny.Tests/JesterTests.cs:60]: in the canHazFunnySln your missing 
Nullability for all projects ❌
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ❌
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ❌
which are all done in canHazFunny.csproj

File: Assignment3/CanHazFunny/OutputJoke.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+namespace CanHazFunny;
+
+public class OutputJoke : IOutputJoke
+{
+    public void Print(string output)
+    {
+        Console.WriteLine(output);
+    }
+}
```
Comment (Bhopss152) [inline Assignment3/CanHazFunny/OutputJoke.cs:11]: add call to start program inside program.cs

File: Assignment3/CanHazFunny.Tests/JesterTests.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            OutputJoke outputJoke = new OutputJoke();
+            JokeService jokeService = new JokeService();
+            Assert.IsNotNull(outputJoke);
+            Assert.IsNotNull(jokeService);
+
+            Jester jester = new Jester(jokeService, outputJoke);
+            Assert.IsNotNull(jester);
+            Assert.AreEqual(jokeService, jester._jokeService);
+            
+        }
```
Comment (Bhopss152) [inline Assignment3/CanHazFunny.Tests/JesterTests.cs:21]: for code readers ease adding //arrange //act //assert just for quick reference not necessary though.

File: Assignment3/CanHazFunny.Tests/CanHazFunny.Tests.csproj [hunk 4]
```diff
@@ -2,7 +2,11 @@
 
   <PropertyGroup>
     <TargetFramework>net6.0</TargetFramework>
-
+    <LangVersion>11.0</LangVersion>
+	  <Nullable>enable</Nullable>
+	  <EnableNETAnalyzers>true</EnableNETAnalyzers>
+	  <CodeAnalysisTreatWarningsAsErrors>true</CodeAnalysisTreatWarningsAsErrors>
+	  <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
```
Comment (mmwoodfo) [inline Assignment3/CanHazFunny.Tests/CanHazFunny.Tests.csproj:9]: You've got some indentation that doesn't need to be here.

Comment (mmwoodfo) [reply Assignment3/CanHazFunny.Tests/CanHazFunny.Tests.csproj:9]: Also, make sure you include these in your test project!

File: Assignment3/CanHazFunny/Jester.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
+﻿using System;
+
+namespace CanHazFunny;
+
+public class Jester
+{
+
+    public IJokeService JokeService { get; }
+    public IOutputJoke OutputJoke { get; }
+    public string? Joke { get; set; }
```
Comment (mmwoodfo) [inline Assignment3/CanHazFunny/Jester.cs:10]:
```suggestion

```
You should keep this scoped to the method it's used in, there isn't a reason you'd want to access outside the TellJoke method.

File: Assignment3/CanHazFunny/Jester.cs [hunk 6]
```diff
@@ -0,0 +1,26 @@
+﻿using System;
+
+namespace CanHazFunny;
+
+public class Jester
+{
+
+    public IJokeService JokeService { get; }
+    public IOutputJoke OutputJoke { get; }
```
Comment (mmwoodfo) [inline Assignment3/CanHazFunny/Jester.cs:9]:
```suggestion
    private IJokeService _JokeService;
    private IOutputJoke _OutputJoke;
```
If these don't need to be public, make them private.

File: Assignment3/CanHazFunny/Jester.cs [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    {
+        JokeService = jokeService ?? throw new ArgumentNullException(nameof(jokeService));
+        OutputJoke = outputJoke ?? throw new ArgumentNullException(nameof(outputJoke));
+    }
+  
+    public void TellJoke()
+    {
+        do
+        {
+            Joke = JokeService.GetJoke();
```
Comment (mmwoodfo) [inline Assignment3/CanHazFunny/Jester.cs:21]: `ctrl+K,D` to format your code!

**Pull request closed without merging**